### PR TITLE
ci: add auto version tracking + build/test/promote pipeline with real-c8y integration test

### PR DESCRIPTION
**Pull request by mstoffel-sag**
## Summary

Adds an end-to-end CI pipeline for the OPC-UA device gateway container, plus automatic upstream version tracking.

### What runs and when

- **Daily cron + manual dispatch** (`.github/workflows/check-versions.yaml`):
  Compares the latest OPC-UA installer version on the public download page and the latest thin-edge `tedge-release` Cloudsmith package against `.github/tracked-versions.json`. If anything changed, it commits to the rolling branch `bot/bump-versions` and opens/updates a PR. Uses `WORKFLOW_PAT` so the PR triggers downstream `pull_request` workflows (the default `GITHUB_TOKEN` would not).

- **Build / integration test on every push and same-repo PR** (`.github/workflows/publish.yaml`, job `build-and-test`):
  1. Builds amd64 and pushes the freshly built image to ghcr.io under a throwaway tag `ci-<run_id>-<run_attempt>`.
  2. Runs the reusable composite action `.github/actions/integration-test` against a real Cumulocity tenant (`C8Y_HOST` / `C8Y_USER` / `C8Y_PASSWORD` secrets):
     - Downloads upstream `startOPCDemoContainer.sh` and `docker-compose-opcua-device-gateway-demo-container.yml`.
     - Patches the gateway compose `image:` to point at the throwaway ghcr.io tag.
     - Suffixes the two shared software names (`opcua-server`, `opcua-device-gateway`) with the run-scoped `DEVICE_NAME = test-opcua-<run_id>` so concurrent runs cannot collide on the c8y software repository.
     - Runs the upstream demo script verbatim against the real tenant.
     - Verifies the `Pump01` device of type `c8y_OpcuaDevice` was created.
     - Cleans up: deletes the demo device + its child devices + the trusted certificate, stops the local compose stack, and deletes the run-scoped software entries.
  3. Always-runs cleanup: deletes the throwaway ghcr.io `ci-…` tag.

- **Multi-arch publish on push to `main` and on dispatch only** (`publish.yaml`, job `publish`):
  Builds `linux/amd64,linux/arm64,linux/arm/v7` and pushes three tags to ghcr.io:
  - `<opcua>-tedge<tedge>` (rolling),
  - `<opcua>-tedge<tedge>-YYYYMMDD-<short-sha>` (immutable, lets you pin a known-good build even when the rolling tag rolls forward later in the same day),
  - `latest` (rolling).

- **Manual integration retest** (`.github/workflows/test-integration.yaml`):
  Standalone `workflow_dispatch` that runs the same composite action against an arbitrary image ref (e.g. `ghcr.io/.../opcua-device-gateway:latest`). Useful for re-validating an already-published image without rebuilding.

### Cross-repo / fork-PR safety

`build-and-test` is gated with `head.repo.full_name == github.repository`, so:
- Same-repo PRs (including the bot's `bot/bump-versions`) → full pipeline runs.
- External fork PRs → job is skipped (secrets not available, can't push to ghcr.io anyway).

### Why ghcr.io for the CI tag (not a local registry)

The first iteration used a local `registry:2` service container on the runner. That fails because podman runs **inside** the gateway container under test, where `localhost:5000` resolves to the container itself, not the runner host. Pushing the CI image to ghcr.io with a throwaway tag is the simplest path that keeps the unmodified gateway image as the test subject. The temp tag is deleted in an `always()` cleanup step.

### Other concerns from review

- `paths-ignore: ['**/*.md', 'LICENSE', 'docs/**']` on both `push` and `pull_request` triggers prevents doc-only commits from rebuilding/republishing the image.
- Per-PR concurrency cancels superseded runs; the rest of the cron / dispatch traffic is serialized via `opcua-gateway-ci`.
- Each `workflow_dispatch` input (`opcua_version`, `tedge_version`) falls back to `tracked-versions.json` independently.
- The cron uses `tedge-release` (not `tedge-main`) and does not strip metadata from the version string.

### Validation

The pipeline was tested end-to-end in `mstoffel-sag/opcua-device-gateway-container#3`. Latest run: build → push to ghcr.io ci tag → integration test against the live tenant → Pump01 verification → cleanup → ci tag deletion all green.

### New / changed files

- `.github/workflows/check-versions.yaml`
- `.github/workflows/publish.yaml`
- `.github/workflows/test-integration.yaml`
- `.github/actions/integration-test/action.yml`
- `.github/tracked-versions.json`
- `containers/opcua-device-gateway/Dockerfile` (accepts `TEDGE_VERSION` build-arg)
- `containers/opcua-device-gateway/entrypoint.sh` (c8y.http fallback)

### Required secrets

- `C8Y_HOST`, `C8Y_USER`, `C8Y_PASSWORD` — Cumulocity tenant for integration tests.
- `WORKFLOW_PAT` — repo-scoped PAT used by `check-versions.yaml` so its bot PRs trigger downstream workflows.

### CODE REVIEW

File: .github/workflows/check-versions.yaml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        run: |
+          # Query Cloudsmith API for the latest tedge Debian package (amd64) from tedge-main
+          FULL=$(curl -sf \
+            "https://api.cloudsmith.io/v1/packages/thinedge/tedge-main/?q=name%3Atedge+format%3Adeb+architecture%3Aamd64&page_size=10&sort=-date" \
+            | jq -r '[.[] | select(.name == "tedge")] | .[0].version')
+          if [ -z "$FULL" ] || [ "$FULL" = "null" ]; then
+            echo "::error::Failed to determine latest thinEdge version"
+            exit 1
+          fi
+          # Strip Debian build metadata — keep only the leading semver (e.g. "2.0.1~109+g36b9024" -> "2.0.1")
```
Comment (reubenmiller) [inline .github/workflows/check-versions.yaml]: You shouldn't be stripped the meta information here, as it gives you the wrong version. e.g. `2.0.1` doesn't actually exist.

Instead just use the `tedge-release` instead of `tedge-main`, as the former is the repo for the official releases (e.g. the non-auto generated numbering).

However it would make sense to also test against the most recent thin-edge.io version on the main branch (from the `tedge-main` repo), as this would be able to spot errors before the official release is done.

File: .github/workflows/check-versions.yaml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+          cat .github/tracked-versions.json
+
+      - name: Commit and push updated versions
+        if: steps.changed.outputs.changed == 'true'
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "github-actions[bot]@users.noreply.github.com"
+          git add .github/tracked-versions.json
+          git commit -m "chore: bump versions [opcua=${{ steps.opcua.outputs.version }}, tedge=${{ steps.tedge.outputs.version }}]"
+          git push
```
Comment (reubenmiller) [inline .github/workflows/check-versions.yaml]: If the workflow is automatically creating a commit (on the main branch?), then I assume you want the commit to also trigger another workflow? If so, then you'll need to use a Personal Access Token instead of GITHUB_TOKEN, as the latter doesn't trigger downstream workflows (as this is a protection mechanism by Github).

File: .github/workflows/check-versions.yaml (hunk 1 above)
Comment (reubenmiller) [reply .github/workflows/check-versions.yaml]: Or is this meant to create a PR? I think the PR would be preferred as you don't want to update the versions before knowing if they actually work or not.

**Comment by reubenmiller** [comment]
There are some flaws in the PR

* We shouldn't be committing directly to main, but rather create a PR which then runs the pre-checks with the newly proposed tracking versions...then once the PR checks are done, then the stuf
* We shouldn't publish the images to the ghcr.io which are just used for a temporary basis. Instead either use docker-compose build functionality, or publish to a local registry on the CI runner, and reference them from the docker-compose file in the tests. See https://docs.docker.com/build/ci/github-actions/local-registry/
* If a commit, or PR is created by a workflow, then it needs to use a PAT, otherwise any workflows that are normally triggered won't be (due to Github protection mechanisms)

And other non factual points:

> A small but important detail: go-c8y-cli v2 silently no-ops write commands in non-TTY environments unless C8Y_SETTINGS_SESSION_MODE=dev AND C8Y_SETTINGS_DEFAULTS_ALLOWEMPTYPIPE=true. Both are set in test-integration.yaml.

This isn't true / it is a hallucination. You just need to use `-n` on all commands which don't expected piped input. The allow empty pipe is used for something else and not related to non-tty environments...and technically it has nothing to do with non-tty environment and a general problem cause by github mapping stdin to a FIFO buffer which affects the go-c8y-cli piped input detection.

File: .github/workflows/check-versions.yaml (hunk 1 above)
Comment (mstoffel-sag) [reply .github/workflows/check-versions.yaml]: It only tags the image  if the test is successful.

Comment (reubenmiller) [reply .github/workflows/check-versions.yaml]: Though ideally we shouldn't be committing to the main branch for unproven things (even if it is for tracking).

**Comment by mstoffel-sag** [comment]
Update — addressed remaining feedback and validated end-to-end:

- Switched the integration-test image away from a local `registry:2` to a throwaway ghcr.io tag `ci-<run_id>-<run_attempt>`. The local-registry approach failed because podman runs inside the gateway container under test, where `localhost:5000` resolves to the container, not the runner. The temp tag is deleted in an `always()` cleanup step.
- Suffixed `opcua-server` and `opcua-device-gateway` software names with the run-scoped `DEVICE_NAME` so concurrent CI runs cannot collide on the c8y software repository, and cleanup deletes only what this run created.
- `build-and-test` job is gated with `head.repo.full_name == github.repository`, so external fork PRs (which can't access secrets / push to ghcr.io) skip it cleanly.
- PR description above has been rewritten to reflect the final design.

Validation run (mstoffel-sag fork): build → ghcr.io ci tag push → integration test against live tenant → Pump01 verified → cleanup → ci tag deleted, all green.
